components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/DefaultTemplateManager.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright (c) 2024, WSO2 LLC. (http://www.wso2.com).
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.email.mgt.store;
+
+import org.wso2.carbon.identity.governance.exceptions.notiification.NotificationTemplateManagerServerException;
+import org.wso2.carbon.identity.governance.model.NotificationTemplate;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This class serves as a unified template management system that delegates the template persistence operations
+ * to both template persistent manger crafted from the factory  and an in-memory manager.
+ * This class will function as a wrapper class for the template manager produced from the factory.
+ */
+public class DefaultTemplateManager implements TemplatePersistenceManager {
+
+    private final TemplatePersistenceManager templatePersistenceManager;
+    private final TemplatePersistenceManager inMemoryTemplateManager = new InMemoryBasedTemplateManager();
+
+    public DefaultTemplateManager() {
+
+        TemplatePersistenceManagerFactory templatePersistenceManagerFactory = new TemplatePersistenceManagerFactory();
+        this.templatePersistenceManager = templatePersistenceManagerFactory.getTemplatePersistenceManager();
+    }
+
+    @Override
+    public void addNotificationTemplateType(String displayName, String notificationChannel, String tenantDomain)
+            throws NotificationTemplateManagerServerException {
+
+        templatePersistenceManager.addNotificationTemplateType(displayName, notificationChannel, tenantDomain);
+    }
+
+    @Override
+    public boolean isNotificationTemplateTypeExists(String displayName, String notificationChannel, String tenantDomain)
+            throws NotificationTemplateManagerServerException {
+
+        return templatePersistenceManager.isNotificationTemplateTypeExists(displayName, notificationChannel,
+                tenantDomain) ||
+                inMemoryTemplateManager.isNotificationTemplateTypeExists(displayName, notificationChannel,
```

Comment:
Lets have inMemoryTemplateManager as the left operand to improve performance, since 'OR' operator's associativity is left to right. 

Basically this will allows to check existence for a template types, using in memory map first and then do a db call only when needed.

Reply:
Fixed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/256/commits/8628c42d5f74b1e58a734be7fc641c3f846af46d